Implementing the OAuth2 Device Authorization Grant

justin-slowik opened the pull request:
Fixes #911

This PR adds functionality for the [OAuth 2 Device Flow](https://oauth.net/2/device-flow/)

I initially wrote this a few months back for my company's private fork, but have rebased it and would like to submit if possible.

Basic usage:
* Device sends a POST to the /device/code endpoint with the client id/secret, and scopes.  Dex returns a device code, user code, expire time, query interval, verification uri and "complete" verification uri (pre-populates the verification page with the user code)
`curl https://127.0.0.1/dex/device/code -d client_id=platform_device -d scope="openid profile email offline_access"
`
> {
   "device_code": "powkm5zcpz7njeoyh63fminzxrcb5ystrjocjx5ch7s47ezmpli",
   "user_code": "FQXZ-HRCG",
   "verification_uri": "http://127.0.0.1/dex/device",
   "verification_uri_complete": "http://127.0.0.1/dex/device?user_code=FQXZ-HRCG",
   "expires_in": 300,
   "interval": 5
}

* The device periodically (at the rate defined by "interval") queries the /device/token endpoint, including the device_code in the POST body.  If the end user has not gone to the device_verification_uri and logged in yet, the error "authorization_pending" will be returned.  If the /device/token endpoint is queried with the device code faster than a rate defined by "interval", a "slow_down" message is retuned, and the wait interval is increased by an additional 5 seconds (from 5 seconds, to 10 seconds, to 15, etc)

* As the device queries the /device/token endpoint, the user can go to the verification URI, enter their verification code, then log in using the standard authorization code flow.  Upon successful login, the user will be alerted to return to their device.

* After the user has logged in, when the advice hits the /device/token endpoint, the access_token will be returned.

Just a note:   rate limiting is only included for the /device/token endpoint.  For our use case, all other rate limiting is currently handled at the API gateway.  

I do not have a large amount of time to do additional work on this branch due to other commitments (other than possibly fixing showstopping issues), so people are welcome to take over.

bonifaido commented:
Great and really much-awaited feature! I have added some reviewers to be able to get this into master.

justin-slowik commented on examples/config-dev.yaml:
```diff
@@ -65,7 +66,10 @@ staticClients:
   name: 'Example App'
   secret: ZXhhbXBsZS1hcHAtc2VjcmV0
-
+#  - id: example-device-client
+#    redirectURIs:
+#      - /dex/device/callback
+#    name: 'Static Client for Device Flow'
```
For the device flow to work properly with redirect URIs, I set up my device clients to register its device callback as its valid redirect URI

justin-slowik commented:
> Great and really much-awaited feature! I have added some reviewers to be able to get this into master.

Glad to help.  We've been running testing it in-house for a few months now and nothing gamebreaking has come up from a functionality standpoint.  

The only thing I'm iffy on is having to define "/device/callback" as a redirect uri for any client you want to leverage for the device flow.  I considered hardcoding it, but considered the use case where clients may be defined that didn't want external devices receiving access tokens.

JoelSpeed commented on storage/sql/migrate.go:
```diff
@@ -229,4 +229,25 @@ var migrations = []migration{
+				device_code text not null,
+				client_id text not null,
+				client_secret text ,
+				scopes bytea not null, -- JSON array of strings
```
Is this kind of comment allowed in SQL statements like this? Never seen it before 😅

justin-slowik replied:
@JoelSpeed  i actually copied this comment format from earlier in migrate.go.  it was used in creating the older tables :) (see client, auth_request, auth_code, refresh_token, and keys )

JoelSpeed commented on storage/storage.go:
```diff
@@ -24,9 +25,21 @@ var (
+}
+
+func newSecureID(len int) string {
+	buff := make([]byte, len) // 128 bit random ID.
```
The comment on this line is no longer correct

justin-slowik replied:
that it is not, ill tweak it :)

JoelSpeed commented on storage/storage.go:
```diff
@@ -101,8 +120,9 @@ type Storage interface {
+	UpdateDeviceToken(deviceCode string, updater func(t DeviceToken) (DeviceToken, error)) error
 
-	// GarbageCollect deletes all expired AuthCodes and AuthRequests.
+	// GarbageCollect deletes all expired AuthCodes,AuthRequests, DeviceRequests, and DeviceTokens.
```
```suggestion
	// GarbageCollect deletes all expired AuthCodes, AuthRequests, DeviceRequests and DeviceTokens.
```

justin-slowik commented on server/deviceflowhandlers.go:
```diff
@@ -0,0 +1,386 @@
+	switch r.Method {
+	case http.MethodGet:
+		userCode := r.URL.Query().Get("user_code")
+		invalidAttempt, err := strconv.ParseBool(r.URL.Query().Get("invalid"))
+		if err != nil {
+			invalidAttempt = false
```
I
![Screen Shot 2020-06-02 at 2 24 13 PM](https://user-images.githubusercontent.com/54368740/83555784-d023f800-a4dc-11ea-8c67-747a2e7f9a54.png)
f the user inputs an incorrect/invalid device code, this set a basic flag so that a user can be alerted on the page and try again.

JoelSpeed replied:
Maybe this is worth a comment to explain what invalidAttempt achieves? I think I'd forget if I came back to this in 6 months

justin-slowik replied:
@JoelSpeed  fair enough, will do!

justin-slowik commented on server/templates.go:
```diff
@@ -28,14 +30,17 @@ var requiredTmpls = []string{
 	tmplPassword,
 	tmplOOB,
 	tmplError,
+	tmplDevice,
```
good catch, my mistake

justin-slowik commented on storage/conformance/conformance.go:
```diff
@@ -881,3 +964,72 @@ func testTimezones(t *testing.T, s storage.Storage) {
+
+	//Validate expected result set
+	if got.Status != "complete" {
+		t.Fatalf("update failed, wanted token status=%#v got %#v", "complete", got.Status)
```
I honestly do not remember when I added moved from %v to %#v.  i see no real reason for it.. i'll edit it

JoelSpeed commented on server/server_test.go:
```diff
@@ -1253,3 +1276,160 @@ func TestRefreshTokenFlow(t *testing.T) {
+			data := url.Values{}
+			data.Set("client_id", clientID)
+			data.Add("scope", strings.Join(requestedScopes, " "))
+			//for _, scope := range requestedScopes {
+			//	data.Add("scope", scope)
+			//}
```
Can we remove this?

justin-slowik replied:
of course, didn't realize I left in commented code :)

wolfeidau commented:
This is really cool and would love to see this merged/completed!

Is there anything I can do to help out?

justin-slowik commented:
@JoelSpeed  believe I hit all the changes you asked for then (after some accidental rebase chaos) got everything back to normal again. :)

bonifaido commented on server/deviceflowhandlers_test.go:
```diff
@@ -0,0 +1,672 @@
+package server
```
This test is 👏 !

JoelSpeed commented on storage/conformance/conformance.go:
```diff
@@ -834,6 +836,87 @@ func testGC(t *testing.T, s storage.Storage) {
+			t.Errorf("garbage collection failed: %v", err)
+		} else {
+			if result.DeviceRequests != 0 {
+				t.Errorf("expected no device garbage collection results, got %#v", result)
```
Do we want the `%#v` here?

justin-slowik replied:
@JoelSpeed I was following the pattern that was done above for authRequests and AuthCodes (see lines 768, 820).  No issue with changing it, but was trying to conform to existing patterns :)

JoelSpeed commented on storage/conformance/conformance.go:
```diff
@@ -881,3 +964,72 @@ func testTimezones(t *testing.T, s storage.Storage) {
+		t.Fatalf("update failed, wanted token status=%v got %v", "complete", got.Status)
+	}
+	if got.Token != "token data" {
+		t.Fatalf("update failed, wanted token =%v got %v", "token data", got.Token)
```
Extra `=`?
```suggestion
		t.Fatalf("update failed, wanted token %v got %v", "token data", got.Token)
```

JoelSpeed commented on server/deviceflowhandlers_test.go:
```diff
@@ -0,0 +1,672 @@
+		scopes                 []string
+		expectedResponseCode   int
+		expectedServerResponse string
+	}{
+		{
+			testName:             "New Valid Code",
+			clientID:             "test",
+			scopes:               []string{"openid", "profile", "email"},
+			expectedResponseCode: http.StatusOK,
+		},
+	}
```
Could we add some more cases here? Eg, what happens with an invalid code? `expectedServerResponse` is currently unused?

justin-slowik replied:
The name is a bit misleading.  All this test does is hit the /device/code endpoint, and validate we can get a device code response (device code, user code, verfication uri, expire time, and request limit) in the proper format.  

Ive updated the tests to include a test case for using an invalid request type (GET vs POST) for code coverage, completeness though

JoelSpeed commented on server/deviceflowhandlers_test.go:
```diff
@@ -0,0 +1,672 @@
+			})
+			defer httpServer.Close()
+
+			if err := s.storage.CreateAuthCode(tc.testAuthCode); err != nil {
+				t.Errorf("failed to create auth code: %v", err)
+			}
+
+			if err := s.storage.CreateDeviceRequest(tc.testDeviceRequest); err != nil {
+				t.Errorf("failed to create device request: %v", err)
+			}
+
+			if err := s.storage.CreateDeviceToken(tc.testDeviceToken); err != nil {
+				t.Errorf("failed to create device token: %v", err)
+			}
```
Should these be fatal errors for the test? I don't think the rest of the test is valid if these don't succeed?

justin-slowik replied:
Done!

wolfeidau commented:
Any movement on this?

I would love to road test this.

justin-slowik commented:
> Any movement on this?
> 
> I would love to road test this.

@wolfeidau I believe I've addressed all comments received.  @JoelSpeed  anything else you need on my end?

JoelSpeed reviewed (approved):
All good, let's merge!
